### ISSUE TITLE
Every other infremen task is being dropped (general dropping of tasks in schedule)

### ISSUE THREAD

**Issue by marc-hanheide** [strands-project/aaf_deployment#337]
this is due to the only 7 seconds between the end of one and the beginning of the next, giving the robot only 7 seconds to reach the next place. If our travel estimate is bigger than this, then the task will be dropped. Talking to @hawesie it seems we have merely two options here:
1. increasing the window size, e.g. setting `end_before` to give more room before the next task starts
2. reducing `max_duration`, which conceptually might be better, but will remove the guarantee that the task starts at `start_after`. 

In any case, we need to give the scheduler more wriggle room to be able to schedule the tasks, including the travel time. So, a good idea might be to take the average travel times and subtract them from `max_duration`... Up for discussion between @gestom @hawesie @Jailander @cdondrup 

We (@hawesie and I) have discussed that an "opportunistic" execution would be better, where a task is only interrupted after `max_duration` if another task is scheduled, rather than always just cancel it even if there's no need.

**Comment by marc-hanheide** [comment]
_Here's the trail of messages of the discussion_

## Christian Dondrup _2016-04-15T09:42:29.526Z_:

I'll test the PR after lunch.

---

## Nick Hawes _2016-04-15T09:42:38.521Z_:

https://github.com/strands-project/strands_executive/pull/226

---

## Christian Dondrup _2016-04-15T13:02:17.083Z_:

Started to test things. Robot won't move currently.

---

## Marc Hanheide _2016-04-15T13:33:07.460Z_:

cool, it publishes the drop events now!

---

## Marc Hanheide _2016-04-15T13:33:20.852Z_:

nice

---

## Christian Dondrup _2016-04-15T13:34:10.681Z_:

I'll leave it running like this for now. The navigation PR that Jaime just opened is also in there.

---

## Marc Hanheide _2016-04-15T13:34:29.778Z_:

:+1: 

---

## Christian Dondrup _2016-04-15T13:34:44.006Z_:

I tested the info_terminal_start task again. Works. Let's see on monday if it really does

---

## Marc Hanheide _2016-04-15T13:41:48.789Z_:

```
15/04/2016, 15:40:00    15 [TASK_FAILED]        5   /info_task_server   Teekueche   15/04/2016, 15:40:05    15/04/2016, 15:40:05    15/04/2016, 15:44:58    1
15/04/2016, 15:40:00    6 [NAVIGATION_FAILED]       5   /info_task_server   Teekueche   15/04/2016, 15:40:05    15/04/2016, 15:40:05    15/04/2016, 15:44:58    1
15/04/2016, 15:39:55    13 [DROPPED]    Dropped because it is now not possible to be executed in the time available 6   /info_task_server   Teekueche   15/04/2016, 15:45:05    15/04/2016, 15:45:05    15/04/2016, 15:49:58    1
15/04/2016, 15:39:55    1 [ADDED]       9   /info_task_server   Frisoer 15/04/2016, 16:00:05    01/01/1970, 01:00:00    15/04/2016, 16:04:58    1
15/04/2016, 15:36:49    4 [NAVIGATION_STARTED]      5   /info_task_server   Teekueche   15/04/2016, 15:40:05    15/04/2016, 15:40:05    15/04/2016, 15:44:58    1
15/04/2016, 15:36:45    3 [TASK_STARTED]        5   /info_task_server   Teekueche   15/04/2016, 15:40:05    15/04/2016, 15:40:05    15/04/2016, 15:44:58    1
15/04/2016, 15:36:14    15 [TASK_FAILED]        3   /info_task_server   WayPoint15  15/04/2016, 15:30:05    15/04/2016, 15:30:05    15/04/2016, 15:34:58    1
15/04/2016, 15:36:14    10 [EXECUTION_FAILED]       3   /info_task_server   WayPoint15  15/04/2016, 15:30:05    15/04/2016, 15:30:05    15/04/2016, 15:34:58    1
```

---

## Marc Hanheide _2016-04-15T13:43:11.404Z_:

@hawesie this is really strange. It’s good we see tasks being dropped now, but that it is dropped is surprising

---

## Nick Hawes _2016-04-15T14:10:23.945Z_:

why is it surprising?  what is the configured max duration?

---

## Nick Hawes _2016-04-15T14:12:27.792Z_:

It was dropped at 15:39:55 it looks like it must complete by  15:49:58 that means the robot must be able to navigate to the target location and run for max_duration in the 10 munutes available to it. was that possible?

---

## Nick Hawes _2016-04-15T14:12:44.299Z_:

(I'm not ruling out bugs, just checking!)

---

## Marc Hanheide _2016-04-15T15:15:06.697Z_:

I thought infremen always asks for tasks of 5 minute length, every 10 minutes… but no I found it does the same [`task.end_before - task.start_after`](https://github.com/strands-project/aaf_deployment/blob/indigo-devel/info_terminal/infremen/src/infremen.cpp#L414) as the other components… But why is it then _ever_ executing anything if the travel time is added?

---

## Marc Hanheide _2016-04-15T15:15:15.525Z_:

I don’t get it...

---

## Marc Hanheide _2016-04-15T15:16:08.738Z_:

I think there’s still some confusion about what all those constraints should be

---

## Marc Hanheide _2016-04-15T15:18:19.297Z_:

you be just me, though

---

## Marc Hanheide _2016-04-15T15:20:02.177Z_:

so, here’s an exemplary info terminal task:

```
    task_id: 28
    start_node_id: WayPoint4
    end_node_id: WayPoint4
    action: /info_task_server
    start_after:
      secs: 1460734505
      nsecs: 0
    end_before:
      secs: 1460734798
      nsecs: 0
    max_duration:
      secs: 293
      nsecs: 0
    execution_time:
      secs: 1460734505
      nsecs: 0
    priority: 1
    arguments: []
```

---

## Marc Hanheide _2016-04-15T15:20:29.870Z_:

`end_before-start_after == 293 == max_duraction`

---

## Marc Hanheide _2016-04-15T15:20:49.622Z_:

this holds true for any `info_task_server` tasks 

---

## Marc Hanheide _2016-04-15T15:21:14.576Z_:

so, given we always have to travel between nodes, it should _never_ work, right? And not just sometimes not?

---

## Marc Hanheide _2016-04-15T15:21:36.873Z_:

sorry, if I’m asking stupid questions

---

## Marc Hanheide _2016-04-15T15:34:57.656Z_:

Just to add to the discussion, another lift of events, with one (to me random one) dropped:

```
15/04/2016, 17:30:32    8 [EXECUTION_STARTED]       27  /info_task_server   WayPoint68  15/04/2016, 17:30:05    15/04/2016, 17:30:05    15/04/2016, 17:34:58    1
15/04/2016, 17:30:32    5 [NAVIGATION_SUCCEEDED]        27  /info_task_server   WayPoint68  15/04/2016, 17:30:05    15/04/2016, 17:30:05    15/04/2016, 17:34:58    1
15/04/2016, 17:29:57    4 [NAVIGATION_STARTED]      27  /info_task_server   WayPoint68  15/04/2016, 17:30:05    15/04/2016, 17:30:05    15/04/2016, 17:34:58    1
15/04/2016, 17:29:55    1 [ADDED]       31  /info_task_server   Cafeteria   15/04/2016, 17:50:05    01/01/1970, 01:00:00    15/04/2016, 17:54:58    1
15/04/2016, 17:29:53    3 [TASK_STARTED]        27  /info_task_server   WayPoint68  15/04/2016, 17:30:05    15/04/2016, 17:30:05    15/04/2016, 17:34:58    1
15/04/2016, 17:27:02    15 [TASK_FAILED]        25  /info_task_server   Kindergarten    15/04/2016, 17:20:05    15/04/2016, 17:20:05    15/04/2016, 17:24:58    1
15/04/2016, 17:27:02    10 [EXECUTION_FAILED]       25  /info_task_server   Kindergarten    15/04/2016, 17:20:05    15/04/2016, 17:20:05    15/04/2016, 17:24:58    1
15/04/2016, 17:24:55    1 [ADDED]       30  /info_task_server   Cafeteria   15/04/2016, 17:45:05    01/01/1970, 01:00:00    15/04/2016, 17:49:58    1
15/04/2016, 17:20:25    8 [EXECUTION_STARTED]       25  /info_task_server   Kindergarten    15/04/2016, 17:20:05    15/04/2016, 17:20:05    15/04/2016, 17:24:58    1
15/04/2016, 17:20:25    5 [NAVIGATION_SUCCEEDED]        25  /info_task_server   Kindergarten    15/04/2016, 17:20:05    15/04/2016, 17:20:05    15/04/2016, 17:24:58    1
15/04/2016, 17:19:55    13 [DROPPED]    Dropped because it is now not possible to be executed in the time available 26  /info_task_server   WayPoint57  15/04/2016, 17:25:05    15/04/2016, 17:25:05    15/04/2016, 17:29:58    1
15/04/2016, 17:19:55    1 [ADDED]       29  /info_task_server   Feuerloescher   15/04/2016, 17:40:05    01/01/1970, 01:00:00    15/04/2016, 17:44:58    1
15/04/2016, 17:18:42    4 [NAVIGATION_STARTED]      25  /info_task_server   Kindergarten    15/04/2016, 17:20:05    15/04/2016, 17:20:05    15/04/2016, 17:24:58    1
15/04/2016, 17:18:38    3 [TASK_STARTED]        25  /info_task_server   Kindergarten    15/04/2016, 17:20:05    15/04/2016, 17:20:05    15/04/2016, 17:24:58    1
15/04/2016, 17:16:50    15 [TASK_FAILED]        23  /info_task_server   WayPoint15  15/04/2016, 17:10:05    15/04/2016, 17:10:05    15/04/2016, 17:14:58    1
15/04/2016, 17:16:50    10 [EXECUTION_FAILED]       23  /info_task_server   WayPoint15  15/04/2016, 17:10:05    15/04/2016, 17:10:05    15/04/2016, 17:14:58    1
```

---

## Marc Hanheide _2016-04-15T15:35:23.576Z_:

@hawesie tell me what’s going on ;-) I’m very puzzled

---

## Nick Hawes _2016-04-15T16:00:58.511Z_:

Regarding `end_before-start_after == 293 == max_duraction`: we do allow some wiggle room, so it’s not super strict. Navigation is separate from the execution time, so the task will only be dropped on time if the robot can’t get there in enough time to execute it within the window.

---

## Nick Hawes _2016-04-15T16:01:45.319Z_:

I wouldn’t assume the dropping/scheduler to be totally rational!

---

## Marc Hanheide _2016-04-15T16:02:17.763Z_:

How much si the wiggle room and where is it configured? I didn’t find it...

---

## Marc Hanheide _2016-04-15T16:02:59.629Z_:

And what would be the suggested strategy if we want a task to “fill” the window, but not exceeding it?

---

## Nick Hawes _2016-04-15T16:03:49.069Z_:

But the other thing to be aware of is that  if you are proposing multiple tasks, the above nav time and execution time constraints must be true for at least one ordering of the tasks. If not tasks will start getting dropped to create a schedule where the constraints are satisfied for a subset of the tasks.

---

## Nick Hawes _2016-04-15T16:05:53.330Z_:

I think I was wrong about the wiggle room — it is handled strictly

---

## Marc Hanheide _2016-04-15T16:06:00.478Z_:

yes, so far so expected, but if it would be strict, then it should never work I understand, as nav times would have to be 0, which they aren’t. So there is wiggle room, but it appears unpredictable. And in order to set a task to “fill” a window, we’d have to know that time on creation, which we can’t...

---

## Marc Hanheide _2016-04-15T16:06:23.253Z_:

if there is no wiggle room how could it ever work?

---

## Nick Hawes _2016-04-15T16:06:35.122Z_:

if there is enough time between windows then it will always work — nav time is not considered in the window

---

## Marc Hanheide _2016-04-15T16:07:35.092Z_:

there is always 7 seconds between windows

---

## Marc Hanheide _2016-04-15T16:07:56.715Z_:

hardly ever enough to go to another place

---

## Nick Hawes _2016-04-15T16:07:59.229Z_:

then I would expect to see every other one dropped

---

## Nick Hawes _2016-04-15T16:08:25.510Z_:

to make room — BUT — I don’t think the dropping is done that smartly 

---

## Marc Hanheide _2016-04-15T16:08:58.587Z_:

that actually appears to be happening (roughly)

---

## Nick Hawes _2016-04-15T16:09:26.319Z_:

so the ones are being dropped to make space to satisfy the other constraints. 

---

## Marc Hanheide _2016-04-15T16:09:44.512Z_:

tasks dropped recently: 26,28, 30,32

---

## Marc Hanheide _2016-04-15T16:10:20.613Z_:

OK, then it’s not a bug, but it’s a problem: So how to best fill the window?

---

## Nick Hawes _2016-04-15T16:10:43.047Z_:

I think make that 7 second gap larger

---

## Nick Hawes _2016-04-15T16:10:57.991Z_:

e.g. to the average navigation duration between any pair of nodes

---

## Nick Hawes _2016-04-15T16:11:10.489Z_:

or just arbitrarily until it works

---

## Nick Hawes _2016-04-15T16:11:30.828Z_:

that 7 seconds is how long you’re giving the system to travel between tasks

---

## Marc Hanheide _2016-04-15T16:12:09.486Z_:

Hmm, ok. But if it’s close it would stand somewhere unable to serve...

---

## Marc Hanheide _2016-04-15T16:12:33.607Z_:

OK, so, the `max_duration == window` is _not_ an actual problem?

---

## Nick Hawes _2016-04-15T16:12:55.788Z_:

Not if you’re not submitting any other tasks

---

## Marc Hanheide _2016-04-15T16:13:31.539Z_:

we are… aren’t we?

---

## Nick Hawes _2016-04-15T16:13:50.161Z_:

that have overlapping time windows with the info terminal ones?

---

## Marc Hanheide _2016-04-15T16:14:04.043Z_:

yes, but with high priority

---

## Nick Hawes _2016-04-15T16:14:17.520Z_:

then the lower priority ones will get dropped anyway

---

## Marc Hanheide _2016-04-15T16:14:39.781Z_:

but that’s probably a 2nd issue

---

## Marc Hanheide _2016-04-15T16:15:28.197Z_:

let’s forget about the priorities for now… so, you recommend changing the window to give more wiggle room between tasks

---

## Marc Hanheide _2016-04-15T16:16:21.185Z_:

but wouldn’t reducing the max_duration do the same, increasing that wiggle room?

---

## Nick Hawes _2016-04-15T16:16:34.746Z_:

Yes, they’d have a similar effect

---

## Marc Hanheide _2016-04-15T16:17:01.993Z_:

but it would lead to a behaviour to keep the task running unless it would have to be interrupted to start navigating towards the next?

---

## Nick Hawes _2016-04-15T16:17:12.328Z_:

Either way you need to give the scheduler enough time to fit in the nav duration to the task and the task itself (max_duration) before end_before

---

## Nick Hawes _2016-04-15T16:17:33.700Z_:

interruptions only happen based on max_duration, not on time needed to reach the next task

---

## Marc Hanheide _2016-04-15T16:17:42.098Z_:

Assume we schedule task to start at 18:00, and end before 18:05, and give it max_duration of 4 minutes

---

## Marc Hanheide _2016-04-15T16:17:54.758Z_:

ah, ok, bugger, so even if there is no next task it would still interrupt?

---

## Nick Hawes _2016-04-15T16:18:25.041Z_:

yeah

---

## Marc Hanheide _2016-04-15T16:18:38.327Z_:

:boom: 

---

## Marc Hanheide _2016-04-15T16:18:52.413Z_:

I dare to say: why ;-) ?

---

## Nick Hawes _2016-04-15T16:19:18.988Z_:

because max_duration is the max duration your task will execute for… leaving it beyond that will cause lock ups

---

## Marc Hanheide _2016-04-15T16:19:21.164Z_:

ok, so there is no chance for some opportunistic “do as long the scheduler allows you to?"

---

## Nick Hawes _2016-04-15T16:19:27.002Z_:

maybe

---

## Nick Hawes _2016-04-15T16:19:57.580Z_:

actually I think if you have the non-interruptible flag set then the scheduler will try to kill the task, but won’t be able to

---

## Marc Hanheide _2016-04-15T16:20:12.432Z_:

I don’t see why it would cause lock-ups as long as you still interrupt it only when you need to

---

## Marc Hanheide _2016-04-15T16:20:15.654Z_:

in execution

---

## Nick Hawes _2016-04-15T16:20:47.454Z_:

yeah, that’s a fair point

---

## Nick Hawes _2016-04-15T16:20:59.048Z_:

but it’s probably not the time to be adding requirements

---

## Marc Hanheide _2016-04-15T16:21:22.170Z_:

that’s probably true… but as it is it can’t stay either

---

## Marc Hanheide _2016-04-15T16:21:42.046Z_:

we have the robot staying in every other place without doing stuff for 5 minutes

---

## Marc Hanheide _2016-04-15T16:22:23.013Z_:

I only missed this so far, because I though it would schedule new tasks only every 10 minutes… (which in fact it did accidentally now)

---

## Nick Hawes _2016-04-15T16:22:33.501Z_:

increasing the travel time between nodes beyond 7 seconds should help though

---

## Marc Hanheide _2016-04-15T16:22:42.017Z_:

we can “tweak” the times, obvisouly

---

## Nick Hawes _2016-04-15T16:23:14.767Z_:

you could also set max duration to something short in a much larger window, but set the is not interruptible flag for 5 minutes 

---

## Marc Hanheide _2016-04-15T16:23:27.081Z_:

Well, ok, let’s do that tomorrow… @Jailander @cdondrup can you reconfigure infremen accordingly? 

---

## Marc Hanheide _2016-04-15T16:23:57.974Z_:

no, we _want_ the info task interruptible… as otherwise staff won’t be able to call more important services for 5 minutes

---

## Nick Hawes _2016-04-15T16:24:32.620Z_:

I would start by working out the average? mode? nav time between the nodes you’re working at and using that as the gap between tasks instead of the 7 secs you have now.

---

## Marc Hanheide _2016-04-15T16:24:37.450Z_:

but ideally, I think we should only interrupt tasks when needed, being more opportnistic

---

## Nick Hawes _2016-04-15T16:25:08.990Z_:

speaking of which, Rosie is going to kill us both if I don’t leave soon

---

## Marc Hanheide _2016-04-15T16:25:15.187Z_:

I can’t see any negative effects of that

---

## Marc Hanheide _2016-04-15T16:25:21.865Z_:

I’m far war ;-)

---

## Marc Hanheide _2016-04-15T16:25:37.171Z_:

But I need to leave as well, Alex Hermans will kill me otherwise

---

## Marc Hanheide _2016-04-15T16:25:56.090Z_:

He’s here for ROVINA coding and I have to find a vegan place for him

---

## Nick Hawes _2016-04-15T16:26:23.333Z_:

say hi to him, and good luck with vegan in rome! 

---

## Marc Hanheide _2016-04-15T16:26:29.194Z_:

Thanks anyway! This was helpful, I do undertsand better and the new “dropping” events made this finally obvious

---

## Nick Hawes _2016-04-15T16:27:43.308Z_:

there is another alternative solution … which is to just use on demand tasks and simply demand one action at a time at the right time effectively negating the scheduler

---

## Nick Hawes _2016-04-15T16:28:05.841Z_:

demanded tasks should interrupt other previously demanded tasks

---

## Marc Hanheide _2016-04-15T16:28:16.249Z_:

true, but that would be a much bigger change than changing towards opportunistic execution

---

## Marc Hanheide _2016-04-15T16:28:46.574Z_:

as we make use of the priorities quite a bit to facilitate the different tasks

---

## Nick Hawes _2016-04-15T16:29:09.866Z_:

I’m not sure actually (I think I could implement it quickly) — we’d just need to make sure demanding respects priorities (which I don;t think it does yet)

---

## Marc Hanheide _2016-04-15T16:29:34.319Z_:

no, it doesn’t I think

---

## Nick Hawes _2016-04-15T16:29:45.500Z_:

but that’s a quick fix

---

## Nick Hawes _2016-04-15T16:29:54.901Z_:

right leaving! ping me tomorrow if needed

---

## Marc Hanheide _2016-04-15T16:30:04.285Z_:

ok, byebye :wave: 

---

## Marc Hanheide _2016-04-15T16:30:08.184Z_:

thanks

---

## Marc Hanheide _2016-04-15T16:35:06.022Z_:

OK, @gestom and @Jailander May I suggest you scroll back about 200 lines ;-) Some useful info in here that wants to be cast into an issue for infremen

**Comment by Jailander** [comment]
I think this could be solved by setting infremen tasks to 1 in this case it behave opportunisticly if I don't recall @gestom wrong. I'll take a look at it anyway but hope @gestom sees this too ;)

**Comment by marc-hanheide** [comment]
I don't really get what you're saying there, but I'll trust it makes sense 😉

**Comment by cdondrup** [comment]
I'm not really sure that makes a difference @Jailander. The schedule is generated every day at midnight, so reconfiguring it to only 1 task at a time would just mean that each task is sent to the scheduler only after the last one finished but it would still be dropped for the same reasons... It shouldn't change the actual behaviour of infremen if there is no special implementation for this. It might even make it worse due to the scheduler needing time to schedule that task right before it should start which might lead to not being able to fit it into the window.

**Comment by Jailander** [comment]
Well as far I understand tasks are only sent when the infoterminal task is not active, so if you only send one task there wont be tasks to be dropped in the execution queue but I am not sure. Will need to take a look tomorrow

**Comment by Jailander** [comment]
Lets hope @gestom sees this first and gives a better idea

**Comment by marc-hanheide** [comment]
Right, the issue that tasks are being dropped needs to be addressed. We shall test the task dropping and interplay between info-terminal etc. better.

**Comment by gestom** [comment]
It seems that this is due to the fact that infremen does not consider the travelling times and it proposes locations that are unreachable in 5 minutes. I will add the implementation that we did with Mirek, which considers that:

1)  Read the topological map and forward it to Mirek's planning.
2) Make a plan of experiments at AAF.
